File: seo-wb-backend/app/schemas/finance.py
from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


GroupBy = Literal["day", "week", "month", "year"]
FinancePeriod = Literal["daily", "weekly"]


class MoneyStringModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class SellerFinanceSettingsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    currency: str | None = None
    default_tax_mode: str | None = None
    default_tax_rate: str | float | None = None
    tax_base: str | None = None
    default_packaging_cost: str | float | None = None
    default_labeling_cost: str | float | None = None
    default_shipping_to_warehouse_cost: str | float | None = None
    default_other_unit_cost: str | float | None = None


class ProductFinanceSettingsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int | None = None
    cost_price: str | float = "0"
    cost_currency: str = "RUB"
    packaging_cost: str | float = "0"
    labeling_cost: str | float = "0"
    shipping_to_warehouse_cost: str | float = "0"
    other_unit_cost: str | float = "0"
    tax_mode: str | None = None
    tax_rate: str | float | None = None
    tax_base: str | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    note: str | None = None

    @model_validator(mode="after")
    def validate_dates(self):
        if self.effective_from and self.effective_to and self.effective_to < self.effective_from:
            raise ValueError("effective_to must be greater than or equal to effective_from")
        return self


class ProductFinanceBulkItemRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: int
    cost_price: str | float = "0"


class ProductFinanceBulkSettingsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[ProductFinanceBulkItemRequest]


class ExternalCostRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cost_date: date
    period_from: date | None = None
    period_to: date | None = None
    cost_type: str
    amount: str | float
    currency: str = "RUB"
    allocation_method: str = "BY_REVENUE"
    product_id: int | None = None
    note: str | None = None

    @model_validator(mode="after")
    def validate_dates(self):
        if self.period_from and self.period_to and self.period_to < self.period_from:
            raise ValueError("period_to must be greater than or equal to period_from")
        return self


class FinanceSyncRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    date_from: date
    date_to: date
    period: FinancePeriod = "daily"
    force: bool = False

    @model_validator(mode="after")
    def validate_dates(self):
        if self.date_to < self.date_from:
            raise ValueError("date_to must be greater than or equal to date_from")
        return self


class FinanceAiAnalyzeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    date_from: date
    date_to: date
    group_by: GroupBy = "day"

    @model_validator(mode="after")
    def validate_dates(self):
        if self.date_to < self.date_from:
            raise ValueError("date_to must be greater than or equal to date_from")
        return self


class PaginationMeta(BaseModel):
    page: int
    perPage: int
    total: int


class SellerFinanceSettingsResponse(BaseModel):
    id: int
    sellerId: int
    currency: str
    defaultTaxMode: str
    defaultTaxRate: str
    taxBase: str
    defaultPackagingCost: str
    defaultLabelingCost: str
    defaultShippingToWarehouseCost: str
    defaultOtherUnitCost: str


class ProductFinanceSettingsResponse(BaseModel):
    id: int
    sellerId: int
    productId: int
    costPrice: str
    costCurrency: str
    packagingCost: str
    labelingCost: str
    shippingToWarehouseCost: str
    otherUnitCost: str
    taxMode: str | None
    taxRate: str | None
    taxBase: str | None
    effectiveFrom: str
    effectiveTo: str | None
    note: str | None


class ProductFinanceSettingsListResponse(BaseModel):
    items: list[ProductFinanceSettingsResponse]
    page: int
    perPage: int
    total: int


class ProductFinanceCatalogItemResponse(BaseModel):
    productId: int
    nmId: int
    vendorCode: str | None
    title: str | None
    subjectName: str | None
    brand: str | None
    photoSquareUrl: str | None
    photoBigUrl: str | None
    hasCostSettings: bool
    settingId: int | None
    costPrice: str | None
    costCurrency: str | None
    packagingCost: str | None
    labelingCost: str | None
    shippingToWarehouseCost: str | None
    otherUnitCost: str | None
    taxMode: str | None
    taxRate: str | None
    taxBase: str | None
    effectiveFrom: str | None
    effectiveTo: str | None
    note: str | None


class ProductFinanceCatalogFacetsResponse(BaseModel):
    brands: list[str]
    subjects: list[str]


class ProductFinanceCatalogResponse(BaseModel):
    items: list[ProductFinanceCatalogItemResponse]
    page: int
    perPage: int
    total: int
    facets: ProductFinanceCatalogFacetsResponse


class MissingFinanceProductResponse(BaseModel):
    id: int
    nmId: int
    vendorCode: str | None
    title: str | None


class MissingFinanceProductsResponse(BaseModel):
    items: list[MissingFinanceProductResponse]


class ExternalCostResponse(BaseModel):
    id: int
    sellerId: int
    costDate: str
    periodFrom: str | None
    periodTo: str | None
    costType: str
    amount: str
    currency: str
    allocationMethod: str
    productId: int | None
    note: str | None


class ExternalCostsListResponse(BaseModel):
    items: list[ExternalCostResponse]
    page: int
    perPage: int
    total: int


class ProductSyncResponse(BaseModel):
    status: str
    totalSynced: int
    cursorUpdatedAt: str | None
    cursorNmId: int | None
    batches: int


class ProductSyncStatusResponse(BaseModel):
    status: str
    cursorUpdatedAt: str | None = None
    cursorNmId: int | None = None
    totalSynced: int | None = None
    lastError: str | None = None
    startedAt: str | None = None
    finishedAt: str | None = None


class WbProductResponse(BaseModel):
    id: int
    nmId: int
    imtId: int | None
    vendorCode: str | None
    brand: str | None
    title: str | None
    description: str | None
    subjectId: int | None
    subjectName: str | None
    photoBigUrl: str | None
    photoSquareUrl: str | None
    sizes: list[dict[str, Any]]
    skus: list[str]
    characteristics: list[dict[str, Any]]
    rawData: dict[str, Any]


class WbProductsListResponse(BaseModel):
    items: list[WbProductResponse]
    page: int
    perPage: int
    total: int


class FinanceSyncResponse(BaseModel):
    status: str
    rowsInserted: int
    lastRrdId: int


class FinanceSyncStatusResponse(BaseModel):
    status: str
    lastRrdId: int | None = None
    totalRows: int | None = None
    lastError: str | None = None


class FinanceRawRowResponse(BaseModel):
    id: int
    rrdId: int
    nmId: int | None
    vendorCode: str | None
    rawData: dict[str, Any]


class FinanceRawRowsResponse(BaseModel):
    items: list[FinanceRawRowResponse]


class FinancePeriodResponse(BaseModel):
    dateFrom: str
    dateTo: str


class FinanceSummaryResponse(BaseModel):
    period: FinancePeriodResponse
    grossRevenue: str
    forPay: str
    wbCosts: str
    cogs: str
    externalAllocatedCosts: str
    profitBeforeTax: str
    taxAmount: str
    profitAfterTax: str
    profitMargin: str
    costCompletenessPercent: str
    rowsCount: int
    productsCount: int


class TimelineItemResponse(BaseModel):
    bucket: str
    forPay: str


class TimelineResponse(BaseModel):
    items: list[TimelineItemResponse]


class ProductBreakdownItemResponse(BaseModel):
    productId: int | None
    nmId: int | None
    vendorCode: str | None
    title: str | None
    quantity: str
    grossRevenue: str
    forPay: str
    wbCosts: str
    cogs: str
    externalAllocatedCosts: str
    profitBeforeTax: str
    taxAmount: str
    profitAfterTax: str
    profitMargin: str
    hasCostSettings: bool
    costMeta: dict[str, Any]


class ProductBreakdownResponse(BaseModel):
    items: list[ProductBreakdownItemResponse]
    page: int
    perPage: int
    total: int


class CostBreakdownResponse(BaseModel):
    wbCosts: str
    cogs: str
    externalAllocatedCosts: str


class InsightItemResponse(BaseModel):
    type: str
    level: Literal["info", "warning", "danger"]
    message: str
    affectedMetric: str
    productIds: list[int]
    recommendedAction: str


class InsightsResponse(BaseModel):
    items: list[InsightItemResponse]


class AllocationPreviewItemResponse(BaseModel):
    productId: int | None
    nmId: int | None
    vendorCode: str | None
    allocatedAmount: str


class AllocationPreviewResponse(BaseModel):
    items: list[AllocationPreviewItemResponse]


class AiAnalyzeResponse(BaseModel):
    snapshotId: int
    analysis: dict[str, Any]


class AiSnapshotResponse(BaseModel):
    id: int
    dateFrom: str
    dateTo: str
    aiAnalysis: dict[str, Any] | None


class AiSnapshotsResponse(BaseModel):
    items: list[AiSnapshotResponse]


class ReconciliationResponse(BaseModel):
    warning: str | None
    calculatedSummary: FinanceSummaryResponse
    reportListCount: int
    reportListTotals: dict[str, str]
    differences: dict[str, str]


class CooldownStateResponse(BaseModel):
    sellerId: int | None
    category: str
    host: str
    method: str
    endpoint: str
    retryAfterSeconds: float
    source: str
    headers: dict[str, Any]


class ApiAvailabilityResponse(BaseModel):
    available: bool
    inCooldown: bool
    activeCooldownCount: int
    cooldowns: list[CooldownStateResponse]


class SyncStatusSummaryResponse(BaseModel):
    status: str | None
    lastSuccessfulAt: str | None
    lastFailedAt: str | None
    lastError: str | None


class FinanceSystemStatusResponse(BaseModel):
    contentApi: ApiAvailabilityResponse
    financeApi: ApiAvailabilityResponse
    commonApi: ApiAvailabilityResponse
    sellerInfoApi: ApiAvailabilityResponse
    activeCooldowns: list[CooldownStateResponse]
    lastSuccessfulProductSyncAt: str | None
    lastSuccessfulFinanceSyncAt: str | None
    lastFailedSyncAt: str | None
    lastFailedSyncError: str | None
    geminiConfigured: bool
    hasProductsMissingFinanceSettings: bool
    missingFinanceSettingsCount: int
    hasUnmappedFinanceRows: bool
    unmappedFinanceRowsCount: int
    automationTimezone: str | None = None
    bootstrapStatus: str | None = None
    bootstrapRangeFrom: str | None = None
    bootstrapRangeTo: str | None = None
    bootstrapFinishedAt: str | None = None
    lastSuccessfulDailySyncDate: str | None = None
    lastDailySyncStatus: str | None = None
    lastDailySyncError: str | None = None
    nextScheduledRunAt: str | None = None

File: seo-wb-frontend/lib/types/finance.ts
export interface CooldownInfo {
  sellerId: number;
  category: string;
  host: string;
  method: string;
  endpoint: string;
  retryAfterSeconds: number;
  source: string;
  headers: Record<string, string | null>;
}

export interface ApiStatusBlock {
  available: boolean;
  inCooldown: boolean;
  activeCooldownCount: number;
  cooldowns: CooldownInfo[];
}

export interface FinanceSystemStatus {
  contentApi: ApiStatusBlock;
  financeApi: ApiStatusBlock;
  commonApi: ApiStatusBlock;
  sellerInfoApi: ApiStatusBlock;
  activeCooldowns: CooldownInfo[];
  lastSuccessfulProductSyncAt: string | null;
  lastSuccessfulFinanceSyncAt: string | null;
  lastFailedSyncAt: string | null;
  lastFailedSyncError: string | null;
  geminiConfigured: boolean;
  hasProductsMissingFinanceSettings: boolean;
  missingFinanceSettingsCount: number;
  hasUnmappedFinanceRows: boolean;
  unmappedFinanceRowsCount: number;
  automationTimezone: string | null;
  bootstrapStatus: string | null;
  bootstrapRangeFrom: string | null;
  bootstrapRangeTo: string | null;
  bootstrapFinishedAt: string | null;
  lastSuccessfulDailySyncDate: string | null;
  lastDailySyncStatus: string | null;
  lastDailySyncError: string | null;
  nextScheduledRunAt: string | null;
}

export interface FinancePeriod {
  dateFrom: string;
  dateTo: string;
}

export interface FinanceSummary {
  period: FinancePeriod;
  grossRevenue: string;
  forPay: string;
  wbCosts: string;
  cogs: string;
  externalAllocatedCosts: string;
  profitBeforeTax: string;
  taxAmount: string;
  profitAfterTax: string;
  profitMargin: string;
  costCompletenessPercent: string;
  rowsCount: number;
  productsCount: number;
}

export interface FinanceTimelinePoint {
  bucket: string;
  forPay: string;
  grossRevenue: string;
  profitAfterTax: string;
}

export interface FinanceTimelineResponse {
  items: FinanceTimelinePoint[];
}

export interface ProductCostMeta {
  costPrice: string;
  packagingCost: string;
  labelingCost: string;
  shippingToWarehouseCost: string;
  otherUnitCost: string;
}

export interface FinanceProductBreakdownItem {
  productId: number | null;
  nmId: number | null;
  vendorCode: string | null;
  title: string | null;
  quantity: string;
  grossRevenue: string;
  forPay: string;
  wbCosts: string;
  cogs: string;
  externalAllocatedCosts: string;
  profitBeforeTax: string;
  taxAmount: string;
  profitAfterTax: string;
  profitMargin: string;
  hasCostSettings: boolean;
  costMeta: ProductCostMeta | null;
}

export interface FinanceProductBreakdownResponse {
  items: FinanceProductBreakdownItem[];
  page: number;
  perPage: number;
  total: number;
}

export interface FinanceCostBreakdown {
  wbCosts: string;
  cogs: string;
  externalAllocatedCosts: string;
}

export interface FinanceInsight {
  type: string;
  level: "info" | "warning" | "danger";
  message: string;
  affectedMetric: string;
  productIds: number[];
  recommendedAction: string;
}

export interface FinanceInsightsResponse {
  items: FinanceInsight[];
}

export type SyncStatus = "idle" | "running" | "completed" | "failed" | "rate_limited" | string;

export interface FinanceSyncStatus {
  status: SyncStatus;
  lastRrdId: number | null;
  totalRows: number | null;
  lastError: string | null;
}

export interface FinanceSyncResponse {
  status: SyncStatus;
  rowsInserted: number;
  lastRrdId: number | null;
}

export interface ProductSyncStatus {
  status: SyncStatus;
  cursorUpdatedAt: string | null;
  cursorNmId: number | null;
  totalSynced: number | null;
  lastError: string | null;
  startedAt: string | null;
  finishedAt: string | null;
}

export interface ProductSyncResponse {
  status: SyncStatus;
  totalSynced: number;
  cursorUpdatedAt: string | null;
  cursorNmId: number | null;
  batches: number;
}

export interface ProductFinanceSetting {
  id: number;
  sellerId: number;
  productId: number;
  costPrice: string;
  costCurrency: string;
  packagingCost: string;
  labelingCost: string;
  shippingToWarehouseCost: string;
  otherUnitCost: string;
  taxMode: string;
  taxRate: string;
  taxBase: string;
  effectiveFrom: string | null;
  effectiveTo: string | null;
  note: string | null;
}

export interface ProductFinanceSettingsResponse {
  items: ProductFinanceSetting[];
  page: number;
  perPage: number;
  total: number;
}

export interface ProductFinanceCatalogItem {
  productId: number;
  nmId: number;
  vendorCode: string | null;
  title: string | null;
  subjectName: string | null;
  brand: string | null;
  photoSquareUrl: string | null;
  photoBigUrl: string | null;
  hasCostSettings: boolean;
  settingId: number | null;
  costPrice: string | null;
  costCurrency: string | null;
  packagingCost: string | null;
  labelingCost: string | null;
  shippingToWarehouseCost: string | null;
  otherUnitCost: string | null;
  taxMode: string | null;
  taxRate: string | null;
  taxBase: string | null;
  effectiveFrom: string | null;
  effectiveTo: string | null;
  note: string | null;
}

export interface ProductFinanceCatalogFacets {
  brands: string[];
  subjects: string[];
}

export interface ProductFinanceCatalogResponse {
  items: ProductFinanceCatalogItem[];
  page: number;
  perPage: number;
  total: number;
  facets: ProductFinanceCatalogFacets;
}

export interface SellerFinanceSettings {
  id: number;
  sellerId: number;
  currency: string;
  defaultTaxMode: string;
  defaultTaxRate: string;
  taxBase: string;
  defaultPackagingCost: string;
  defaultLabelingCost: string;
  defaultShippingToWarehouseCost: string;
  defaultOtherUnitCost: string;
}

export interface ExternalCost {
  id: number;
  sellerId: number;
  costDate: string;
  periodFrom: string;
  periodTo: string;
  costType: string;
  amount: string;
  currency: string;
  allocationMethod: string;
  productId: number | null;
  note: string | null;
}

export interface ExternalCostsResponse {
  items: ExternalCost[];
  page: number;
  perPage: number;
  total: number;
}

export interface GeminiAnalysisResult {
  summary: string;
  insights: unknown[];
}

export interface GeminiFinanceAnalysis {
  snapshotId: number;
  analysis: GeminiAnalysisResult;
}

export interface GeminiSnapshotItem {
  id: number;
  dateFrom: string;
  dateTo: string;
  aiAnalysis: {
    summary: string;
  };
}

export interface GeminiSnapshotsResponse {
  items: GeminiSnapshotItem[];
}

export interface MissingSettingsProduct {
  id: number;
  nmId: number;
  vendorCode: string;
  title: string;
}

export interface MissingSettingsResponse {
  items: MissingSettingsProduct[];
}

export interface WbProduct {
  id: number;
  nmId: number | null;
  imtId: number | null;
  vendorCode: string | null;
  brand: string | null;
  title: string | null;
  description: string | null;
  subjectId: number | null;
  subjectName: string | null;
  photoBigUrl: string | null;
  photoSquareUrl: string | null;
  sizes: unknown[];
  skus: string[];
  characteristics: unknown[];
  rawData: Record<string, unknown>;
  wbUpdatedAt: string | null;
}

export interface WbProductsResponse {
  items: WbProduct[];
  page: number;
  perPage: number;
  total: number;
}

export interface ReconciliationReportListTotals {
  retailAmountSum: string;
  forPaySum: string;
  deliveryServiceSum: string;
}

export interface ReconciliationResponse {
  warning: string | null;
  calculatedSummary: FinanceSummary;
  reportListCount: number;
  reportListTotals: ReconciliationReportListTotals;
  differences: ReconciliationReportListTotals;
}

export type GroupBy = "day" | "week" | "month" | "year";

export interface AllocationPreviewItem {
  productId: number;
  nmId: number;
  vendorCode: string;
  allocatedAmount: string;
}

export interface AllocationPreviewResponse {
  items: AllocationPreviewItem[];
}

export interface SettingsImportResponse {
  imported: number;
  errors: Array<{ row: number; error: string }>;
}
